A34-weighted-round-robin.md

```diff
@@ -0,0 +1,111 @@
+
+
+### Related Proposals:
+* [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md).
+
+## Proposal
+
+The proposal is to carry [`load_balancing_weight`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L108) of [`LbEndpoint`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L76) from [`ClusterLoadAssignment`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint.proto#L34) response to `lb_policy` and implement a `weighted_round_robin` policy based on Earliest deadline first scheduling algorithm picker [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling). Each endpoint will get fraction equal to the [`load_balancing_weight`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L108) for the [`LbEndpoint`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L76) divided by the sum of the `load_balancing_weight` of all `LbEndpoint` within the same [`LocalityLbEndpoints`](https://github.com/envoyproxy/envoy/blob/2dcf20f4baf5de71ba1d8afbd76b0681613e13f2/api/envoy/config/endpoint/v3/endpoint_components.proto#L116) of traffic routed to the locality.
+
+### Overview of `weighted_round_robin` policy
+
+`weighted_round_robin` policy is powered by [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) picker. EDF picker maintains a priority queue of `EdfEntry`. On the top of the queue, it’s the entry with lowest `deadline`. `order_offset` is the tie breaker when two entries have same deadline to maintain FIFO order.
+
+Proposed `EdfEntry`
+```
+struct EdfEntry {
+    // primary key for the priority queue. The entry with least deadline is the top of the queue.
+    double deadline;
+    // secondary key for the priority queue. Used as a tiebreaker for same deadline to maintain FIFO order.
+    uint64 order_offset;
+    // `load_balancing_weight` of this endpoint from the latest `ClusterLoadAssignment`.
```

Comment:
I would prefer that this LB policy be implemented in a way that is not xDS-specific.  It should basically just use the endpoint weight attributes passed into it, and it should not care where those weights come from.  That way, it will be possible to also use this policy without xDS, by writing a custom resolver that provides the same endpoint weight attributes.

To say this another way, I think there are two distinct parts to this proposal: (1) changing the xDS code to populate the addresses' endpoint weight attributes based on the EDS data, and (2) implementing the new LB policy that uses the addresses' endpoint weight attributes.  I suggest structuring this document in a way that makes it clear that these are two separate changes.

Reply:
kk. I will update the doc as suggested structure with two parts (1) and (2).